pytential/source.py

```diff
@@ -76,51 +87,32 @@ def ndofs(self) -> int:
 
     @property
     @abstractmethod
-    def real_dtype(self):
+    def real_dtype(self) -> np.dtype[np.floating]:
         """:class:`~numpy.dtype` of real data living on the source geometry."""
 
     @property
     @abstractmethod
-    def complex_dtype(self):
+    def complex_dtype(self) -> np.dtype[np.complexfloating]:
         """:class:`~numpy.dtype` of complex data living on the source geometry."""
 
     @abstractmethod
-    def op_group_features(self, expr: sym.IntG) -> tuple[Hashable, ...]:
+    def op_group_features(self, expr: IntG) -> tuple[Hashable, ...]:
         """
         :arg expr: a subclass of :class:`~pytential.symbolic.primitives.IntG`.
         :returns: a characteristic tuple by which operators that can be
             executed together can be grouped.
         """
 
-    @abstractmethod
     def get_p2p(self,
                 actx: PyOpenCLArrayContext,
                 target_kernels: tuple[Kernel, ...],
-                source_kernels: tuple[Kernel, ...] | None = None) -> P2PBase:
+                source_kernels: tuple[Kernel, ...] | None = None) -> P2P:
         """
-        :returns: a subclass of :class:`~sumpy.p2p.P2PBase` for evaluating
+        :returns: a subclass of :class:`~sumpy.p2p.P2P` for evaluating
             the *target_kernels* and the *source_kernels* on the source geometry.
         """
 
-    def preprocess_optemplate(self,
-                name: Hashable,
-                discretizations: GeometryCollection,
-                expr: T) -> T:
-        """
-        :returns: a processed *expr*, where each
-            :class:`~pytential.symbolic.primitives.IntG` operator has been
-            modified to work with the current source geometry.
-        """
-        return expr
-
-
-class _SumpyP2PMixin:
-
-    def get_p2p(self,
-                actx: PyOpenCLArrayContext,
-                target_kernels: tuple[Kernel, ...],
-                source_kernels: tuple[Kernel, ...] | None = None) -> P2P:
-        @memoize_in(actx, (_SumpyP2PMixin, "p2p"))
```

Comment:
What was the purpose of this `SumpyP2PMixin`? I tried to do some git blaming, but couldn't find any explanation.

Removed because it displeased `pyright` :grin:

Reply:
Very unclear. I looked through the history, and I don't get it either.